### DISCUSSION TITLE
Modifier timings

### DISCUSSION THREAD

**Discussion by ktunprasert** [houmain/keymapper#90] [Q&A] [answered]
I'm wondering if it's possible to adjust the modifier timing? I mainly use this for my work laptop on the go and the problem I run into is for example: `Space{A} >> Ctrl{A}`
Sometimes this would get triggered when I am typing normally leading me to select all by mistake and overwrite what I'm writing with something else.

**Comment by houmain** [comment]
Hi, have you tried:
`Space{250ms A} >> Control{A}`

**Reply by ktunprasert** [reply]
That does seem a bit fiddly with the timings, is there no way to detect KEY_DOWN vs KEY_SEQUENCE? I will have a play around and report back

**Reply by houmain** [reply]
Maybe `250ms Space{A} >> Control{A}` would be what you need? Unfortunately input sequences must not start with a timeout yet.

**Reply by ktunprasert** [reply]
Apologies, I think I phrased the original topic wrong. It's weird behaviour that modifiers are listening to key in a sequence instead of being able to see the `key_pressed` state in combination with something else. A timer shouldn't dictate whether a `Space A` and `Space{A}` behaves similarly right? If I release the `Spacebar` already then it should know that the next key, within timeout or not, should be the key itself rather than the combo.

**Reply by houmain** [reply]
Do you want to trigger `Control{A}` only when `A` is released and `Space` is still hold?

**Reply by houmain** [reply]
Otherwise I still do not understand how you want to distinguish a `+Space +A -Space -A` from a `+Space +A -A -Space` where + means press and - release.

**Reply by ktunprasert** [reply]
Yes sorry that's what I mean. The combo should only work when the modifier is held down, is this something that's already implemented but I haven't read about?

**Reply by ristomatti** [reply]
Have you tried recording key events without Keymapper? On my mechanical keyboard, if I type two keys in quick sequence, it shows as:
```
key1_press > key2_press > key2_release > key1_release
```

The side effect of this is, I've been able to simulate "leader" key style keymaps (the keys are pressed in sequence rather than held at the same time) by using `A{B} >> X`. Initially I tried with `A !100ms B >> X`, but I couldn't get the action to trigger without specifically making a pause between the keys. Using `A{B} >> X`, I can type them in sequence without having to deliberately wait. I believe it's related to the above behavior on  my keyboard. Example "Leader" key mapping from my config:

```perl
Leader = Comma

# Slack
[class="slack"]
    Leader{F}    >> switch_channel
    Leader{1}    >> goto_workspace1
    Leader{2}    >> goto_workspace2
    Leader{H}    >> goto_home
    Leader{A}    >> goto_activity
    Leader{T}    >> goto_threads
```

**Reply by ristomatti** [reply]
Forgot to mention, I was having a similar issue with this mapping:

```perl
[class="jetbrains-webstorm"]
    Shift{Enter} >> End
```

The issue went away by changing it to:

```perl
[class="jetbrains-webstorm"]
    !Control !Meta Shift{Enter} >> End
```

**Reply by ktunprasert** [reply]
@ristomatti that's a great way to use it, though it conflicts with how I view the "Modifier" in general. In QMK, I use the settings called "Permissive Hold" which does have a time out [link](https://github.com/qmk/qmk_firmware/blob/master/docs/tap_hold.md), though if there is a valid combo within the timeout it immediately executes rather than wait for the timeout to finish. This means I can type fast without needing to wait. If the valid combo is not there, it is a passthrough instead (which is the current behaviour). 

The modifier to me is virtual layers triggered by holding down (Layer Toggle on Hold), and for you is triggered by key sequencing (Tap Dancing). I think for your case `A !100ms B` should be the right syntax, but instead of waiting for the timeout it should be able to see if the next immediate sequence exists as a mapping and executes it immediately. This type of behaviour was introduced into `Kanata` and `KMonad` as well but they never work as well as QMK native solution does (at least 2 timeout are required, **very fiddly**). I only wanted to bring this up to see the repo author's opinions on it, I already use this software and I would be very happy to see it as full replacement for QMK.

**Reply by ristomatti** [reply]
@ktunprasert Agree. My primary use case for Keymapper is application specific shortcuts. I doubt if it's possible to reach the level of precision you'll get out of a hardware solution though.

P.S. I also use QMK w/ 5 layers in addition to i3wm hacks. There's always something to "improve"... Getting it all done through Keymapper would make things quite a lot simpler to manage.

**Reply by houmain** [reply]
@ktunprasert 
> The combo should only work when the modifier is held down, is this something that's already implemented but I haven't read about?

No, this is not yet available but I will try to implement it. I am thinking about enhancing Not. So `Space{A !A}` should trigger when `A` is released and `Space` is still hold.
I also intend to change the behaviour of `A !100ms B`. Currently the timeout starts when `A` is released, afterwards it should start when it is pressed. Then `A !A !100ms B` is needed to get the old behaviour.

**Reply by houmain** [reply, marked as answer]
This is available now in the 3.4.0 release!

**Reply by ristomatti** [reply]
Curious to test this one! I hope it can still be compiled for Ubuntu 20.02 by disabling Wayland support. 🤞

**Reply by ktunprasert** [reply]
> This is available now in the 3.4.0 release!

@houmain Amazing I will test this out, thank you so much and happy holidays 🎉

**Reply by ristomatti** [reply]
This is huge!
